blog/2025-01-17-how-are-the-messages-stored/index.md

```diff
@@ -163,19 +163,21 @@ Here's how we store the metadata for different queue types:
   - for messages stored in the per-queue message store, no data is stored in memory
   - for messages stored in the per-vhost message store, there's some metadata in memory
 * Quorum queues
-  - metadata is stored in memory (at least 32 bytes per message, sometimes a bit more, for example when message TTL is used)
+  - metadata is stored in memory (at least 32 bytes per message, sometimes a bit more, for example when [message TTL](https://www.rabbitmq.com/docs/ttl) is used)
 * Streams
   - no message metadata is stored in memory
 
 This basically means that for messages under 4kb stored in classic queues, as well as for streams, regardless of how many messages
 there are in the queue/stream, the memory usage is constant. You will run out disk before you run out of memory (you should
-configure retention/size limits that, to avoid running out disk, but that's a different story).
+configure [retention](https://www.rabbitmq.com/docs/streams#retention)/[length](https://www.rabbitmq.com/docs/maxlength) limits that,
+to avoid running out disk, but that's a different story).
 
 Here's an illustration highlighting the difference between the two different classic queue storage mechanisms. In this test,
 I published 1 million messages of 4000 bytes each first, then deleted the queue and published 1 million messages of 4100 bytes
 each. As you can see, the memory usage was stable in the first phase (small fluctuations notwithstanding), but when publishing
-larger messages, we can see the memory usage grows as well. This is strictly metadata stored in memory - a million 4kb
-messages would take up 4KB of memory to store, while the actual usage still below 400MB.
+larger messages, we can see the memory usage grows as well. This is because 4100 bytes is above the threshold, so these
+messages are stored in the per-vhost message store and the per-vhost message keeps some metadata in memory. A million 4KB
```

Comment:
so these
messages are stored in the per-vhost message store and the per-vhost message keeps some metadata in memory. 
->
so these
messages are stored in the per-vhost message store and the per-vhost message store keeps some metadata in memory.